gadget/install/content.go

```diff
@@ -120,22 +121,65 @@ func writeFilesystemContent(laidOut *gadget.LaidOutStructure, kSnapInfo *KernelS
 	if laidOut.Role() == gadget.SystemData && kSnapInfo != nil && kSnapInfo.NeedsDriversTree {
 		destRoot := mountpoint
 		if kSnapInfo.IsCore {
-			destRoot = filepath.Join(mountpoint, "system-data")
+			// For core we write the changes in _writable_defaults. The
+			// files are then copied from the initramfs by
+			// populate-writable.service on first boot as the directories
+			// they are in are marked as "transitional" in the
+			// writable-paths file. We cannot copy directly to
+			// "system-data" as that would prevent files already in
+			// _writable_defaults in the directories of interest to not
+			// be copied (because the files are not copied if the
+			// directory exists already).
+			destRoot = filepath.Join(mountpoint, "system-data", "_writable_defaults")
 		}
 		destDir := kernel.DriversTreeDir(destRoot, kSnapInfo.Name, kSnapInfo.Revision)
 		logger.Noticef("building drivers tree in %s", destDir)
 
-		kTargetSysMntPt := snap.MinimalSnapContainerPlaceInfo(kSnapInfo.Name, kSnapInfo.Revision)
+		cpi := snap.MinimalSnapContainerPlaceInfo(kSnapInfo.Name, kSnapInfo.Revision)
 		if err := kernelEnsureKernelDriversTree(
 			kernel.MountPoints{
 				Current: kSnapInfo.MountPoint,
-				Target:  kTargetSysMntPt.MountDir(),
+				Target:  cpi.MountDir(),
 			},
 			nil,
 			destDir,
 			&kernel.KernelDriversTreeOptions{KernelInstall: true}); err != nil {
 			return err
 		}
+
+		// Create mount unit to make the kernel snap content available from
+		// the drivers tree.
+		squashfsPath := dirs.StripRootDir(cpi.MountFile())
+		whereDir := dirs.StripRootDir(cpi.MountDir())
+
+		hostFsType, options := systemd.HostFsTypeAndMountOptions("squashfs")
+		mountOptions := &systemd.MountUnitOptions{
+			Lifetime:                 systemd.Persistent,
```

Comment:
I fail to see the big picture. I have seen that we have to have `/usr/lib/modules` and `/usr/lib/firmware` mounted before we switch root. So we will have to fix that.

This PR adds a mount unit, but for the main boot, not from initrd. If we do need to have a matching mount unit in the initrd to match the one in the main boot, I would expect it to be generate by a generator instead of being just saved in `/etc`.

Reply:
As discussed this is the same unit that will be generated for the snap when seeding. We want still the mounts from initramfs, but that is not done here.